linkerd/core/src/main/scala/io/buoyant/linkerd/ActiveTracer.scala

```diff
@@ -0,0 +1,170 @@
+import com.twitter.finagle.naming.buoyant.DstBindingFactory
+import com.twitter.util.{Future, Stopwatch}
+import io.buoyant.linkerd.ActiveTracer.formatDelegation
+import io.buoyant.namer.{DelegateTree, Delegator}
+import io.buoyant.namer.DelegateTree._
+import io.buoyant.router.RoutingFactory.BaseDtab
+
+private[linkerd] case class RouterContext(
+  routerLabel: String,
+  elapsed: Stopwatch.Elapsed,
+  serviceName: String,
+  clientName: String,
+  selectedAddresses: Option[Set[String]],
+  selectedAddress: String,
+  dtabResolution: List[String]
+) {
+  def formatRouterContext: String = {
+    s"""|--- Router: $routerLabel ---
+          |request duration: ${elapsed().inMillis.toString} ms
+          |service name: $serviceName
+          |client name: $clientName
+          |addresses: [${selectedAddresses.getOrElse(Set.empty).mkString(", ")}]
+          |selected address: $selectedAddress
+          |dtab resolution:
+          |${dtabResolution.map("  " + _).mkString(System.lineSeparator)}
+          |""".stripMargin
+  }
+}
+
+private[linkerd] object RouterContextBuilder {
+  val EmptyDelegateTree = DelegateTree.Empty(Path.empty, Dentry.nop)
+
+  def apply(
+    routerLabel: String,
+    elapsed: Stopwatch.Elapsed,
+    pathCtx: Option[Dst.Path],
+    boundCtx: Option[Dst.Bound],
+    endpoint: EndpointAddr,
+    namers: DstBindingFactory.Namer,
+    dtabs: BaseDtab
+  ): Future[RouterContext] = {
+    val serviceName = pathCtx match {
+      case Some(dstPath) => dstPath.path
+      case None => Path.empty
+    }
+
+    val selectedEndpoint = endpoint.addr match {
+      case inetAddr: Address.Inet => inetAddr.addr.toString.stripPrefix("/")
+      case _ => ""
+    }
+
+    val clientPath = boundCtx match {
+      case None => Path.empty
+      case Some(addrSet) => addrSet.name.id match {
+        case p: Path => p
+        case _ => Path.empty
+      }
+    }
+
+    val lbSet = boundCtx match {
+      case None => Future.value(Addr.Neg)
+      case Some(addrSet) => addrSet.name.addr.changes.toFuture
+    }
+
+    val addresses = lbSet.map {
+      case Addr.Bound(a, _) => Some(
+        a.map {
+          case inetAddr: Address.Inet => inetAddr.addr.toString.stripPrefix("/")
+          case _ => ""
+        }
+      )
+      case _ => None
+    }
+
+    val dtreeF = namers.interpreter match {
+      case delegator: Delegator => delegator.delegate(dtabs.dtab(), serviceName)
+        .map(Some(_))
+      case _ => Future.None
+    }
+
+    dtreeF.joinWith(addresses) {
+      case (dTree: Option[DelegateTree[Name.Bound]], addrSet: Option[Set[String]]) =>
+        val tree = formatDelegation(dTree.getOrElse(EmptyDelegateTree), List.empty, clientPath)
+        RouterContext(
+          routerLabel,
+          elapsed,
+          serviceName.show,
+          clientPath.show,
+          addrSet,
+          selectedEndpoint,
+          tree.map(_.toString)
+        )
+    }
+  }
+}
+
+object ActiveTracer {
```

Comment:
Does this need to be separate object from RequestContextBuilder?  Can these methods just go into that object?